state-sync/state-sync-v2/data-streaming-service/src/stream_engine.rs

```diff
@@ -1624,3 +2013,44 @@ fn create_data_notification(
         data_payload,
     })
 }
+
+/// Extracts the number of new versions and target
+/// ledger info for the given client response payload.
+fn extract_new_versions_and_target(
+    client_response_payload: &ResponsePayload,
+) -> Result<(usize, LedgerInfoWithSignatures), Error> {
+    // Extract the number of new versions and the target ledger info
+    let (num_versions, target_ledger_info) = match &client_response_payload {
+        ResponsePayload::NewTransactionsWithProof((
+            transactions_with_proof,
+            target_ledger_info,
+        )) => (
+            transactions_with_proof.transactions.len(),
+            target_ledger_info.clone(),
+        ),
+        ResponsePayload::NewTransactionOutputsWithProof((
+            outputs_with_proof,
+            target_ledger_info,
+        )) => (
+            outputs_with_proof.transactions_and_outputs.len(),
+            target_ledger_info.clone(),
+        ),
+        response_payload => {
+            // TODO(joshlind): eventually we want to notify the data client of the bad response
```

Comment:
Yeah, this only matters if peers are sending bad responses. In such cases, we notify the data client (so the peer's score can be reduced, and they'll eventually be ignored for some period). These checks are spread across the streaming service and the driver (because we verify different properties in these components). 

Part of the reason for not resolving the TODOs yet is because: (i) it'll require a large refactor, e.g., providing access to the data client/scoring mechanism in the streaming service; and (ii) we have an outstanding task to overhaul the scoring system anyway (because it's not great, has false positives, doesn't disconnect from peers, etc.)